### PR TITLE
Add code signing preview page

### PR DESCRIPTION
**Pull request by bryanoltman**
## Status

**HOLD**

Hold until https://github.com/shorebirdtech/shorebird/pull/2152 is released.

## Description

Adds a top-level page explaining our new patch signing feature and demonstrating how to use it.

Screenshot 2024-05-24 at 4 52 11 PM

### CODE REVIEW

**Review by eseidel — APPROVED** [review]

File: astro.config.mjs [hunk 1]
```diff
@@ -111,6 +111,10 @@ export default defineConfig({
           label: 'Teams',
           link: '/teams',
         },
+        {
+          label: 'Code Signing (preview)',
+          link: '/code-signing',
```
Comment (eseidel) [inline astro.config.mjs]:
```suggestion
          link: '/patch-signing',
```

**Review by eseidel — APPROVED** [review]

File: src/content/docs/code-signing.mdx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+material (or signing via a cloud signing service) please contact us, we'd be
+happy to work with you to support such.
+
+## Generate keys
+
+If you do not already have an RSA key pair you'd like to use, you can generate a
+pair with `openssl`:
+
+```
+# Generate a private key
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply src/content/docs/code-signing.mdx]: Done

File: src/content/docs/code-signing.mdx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+You can verify that the patch is properly signed using `shorebird preview`. On
+the first launch, you should see something like the following in your app logs:
+
+If you close and relaunch your app, you should see this message telling you that
+the patch's signature was verified:
+
+If the patch is missing a signature, or fails signature verification for any
+reason, Shorebird will not load it, and will instead use any previously
+installed and verifiable patch (if there is one) or the unpatched release
+version of your app.
```
[the remark this answers is not in the document]
Comment (eseidel) [reply src/content/docs/code-signing.mdx]: Filed https://github.com/shorebirdtech/shorebird/issues/2150

File: src/content/docs/code-signing.mdx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+your app to crash.
+
+If for any reason you were to lose your private key there is no way to create a
+patch for an application containing the corresponding public key. Even Shorebird
+is not able to create a patch for your application without your private key.
+
+## Trade Offs
+
+The primary trade-off of using patch signing is complexity. Shorebird does not
+yet offer automatic key management, so customers need to create and manage their
```
[the remark this answers is not in the document]
Comment (bryanoltman) [reply src/content/docs/code-signing.mdx]: Done.

File: src/content/docs/patch-signing.mdx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+description: How to sign patches
+---
+
+{/* cspell:words outform pubout */}
+
+:::note
+We've shipped this preview of patch signing to unblock certain customer
+evaluations. While we're confident that this preview adds a useful layer of
+security, it does not yet have all the expected integrations with cloud signing,
+keychains, etc. that a final product would expect to have. We also do not yet
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:12]:
```suggestion
keychains, etc, that a final product would expect to have. We also do not yet
```

File: src/content/docs/patch-signing.mdx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+manage signing keys for you or support many key formats or signature types.
+Signed patches are safe and ready for use, but may have rough edges or missing
+conveniences. Any customer should feel welcome to use this preview in
+production, and is encouraged to file feedback for any rough edges you may find.
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:16]:
```suggestion
production, and is encouraged to file feedback for any bumps you may find.
```
Or something else, just so we don't repeat `rough edges`

File: src/content/docs/patch-signing.mdx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+:::
+
+In addition to our default security measures, Shorebird also provides optional
+patch signing.
+
+Patch signing allows developers to cryptographically sign patch updates with
+their own keys. This ensures that no one (including Shorebird) can change the
+content of your application without your private cryptographic keys.
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:24]:
```suggestion
content of your patch without your private cryptographic keys.
```

File: src/content/docs/patch-signing.mdx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+Signing works in two parts. First, `shorebird release` commands can take an
+optional `--public-key-path` argument to embed a your public key in your
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:27]:
```suggestion
optional `--public-key-path` argument to embed a public key in your
```

File: src/content/docs/patch-signing.mdx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+released application. The Shorebird updater will enforce that only patches
+signed with a corresponding private key will be allowed to load for applications
+that include a public key.
+
+Second, when you build your patch with `shorebird patch`, you can pass
+`--private-key-path` to have Shorebird sign your patch with your private key.
+This is required if you created your release with a public key included.
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:34]: We now require the `public key` too

File: src/content/docs/patch-signing.mdx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+There are no required changes to your code and you can add or remove this
+signing requirement at any time by simply making a new release of your
+application.
+
+## Adding patch signing to your application
+
+To start, you will need an RSA key pair. Shorebird has only tested with a
+limited set of signing algorithms at this time. Please contact us if you have
+other requirements—we'd be happy to work with you to support your needs.
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:44]:
```suggestion
other requirements, we'd be happy to work with you to support your needs.
```

File: src/content/docs/patch-signing.mdx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+The `shorebird` tool expects to be able to read private and public keys
+from `.pem` files on disk. If you need other ways of accessing your key material
+(or signing via a cloud signing service) please contact us. We'd be happy to
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:48]: I feel this paragraph kinda steps into what the previous one already said, about we being happy to support other needs.

File: src/content/docs/patch-signing.mdx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+We're using Android to demo, but this will work on any platform Shorebird
+supports.
+:::
+
+### Create a signed patch
+
+To create a signed patch, run the following command:
+
+```sh
+shorebird patch android --private-key-path /path/to/private.pem
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:91]:
```suggestion
shorebird patch android --private-key-path /path/to/private.pem --public-key-path /path/to/public.pem
```

File: src/content/docs/patch-signing.mdx [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+```
+
+This tells shorebird to sign the patch with the private key you provide.
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:94]:
```suggestion
This tells shorebird to sign the patch with the key pair you provided.
```

File: src/content/docs/patch-signing.mdx [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+Storing private keys on disk is not best-practice, and we recognize
+this. This was built as a demonstration of the system, we look forward to
+working with customers to integrate Shorebird to using the signing services and
+key storage methods they are used to.
+:::
+
+### Test it out
+
+You can verify that the patch is properly signed using `shorebird preview`. On
+the first launch, you should see something like the following in your app logs:
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:106]: Are the logs missing?

File: src/content/docs/patch-signing.mdx [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+If you close and relaunch your app, you should see this message telling you that
+the patch's signature was verified:
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:109]: Same here

File: src/content/docs/patch-signing.mdx [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+If for any reason you were to lose your private key, there is no way to create a
+patch for an application containing the corresponding public key. Even Shorebird
+is not able to create a patch for your application without your private key.
+In such a case, you would need to make and distribute a new release of your
+application to send patches to it.
+
+## Trade Offs
+
+The primary trade-off of using patch signing is complexity. Shorebird does not
+yet offer automatic key management, so customers you to create and manage their
```
Comment (erickzanardo) [inline src/content/docs/patch-signing.mdx:137]:
```suggestion
yet offer automatic key management, so customers are to create and manage their
```

File: src/content/docs/patch-signing.mdx (hunk 7 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx]: Done

File: src/content/docs/patch-signing.mdx (hunk 9 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:35]: Good catch, added.

File: src/content/docs/patch-signing.mdx (hunk 5 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:12]: "etc." should be followed by a period as it is an abbreviation.

File: src/content/docs/patch-signing.mdx (hunk 6 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx]: Done

File: src/content/docs/patch-signing.mdx (hunk 8 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx]: Done

File: src/content/docs/patch-signing.mdx (hunk 10 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:45]: An em-dash is more correct here

File: src/content/docs/patch-signing.mdx (hunk 12 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx]: Done

File: src/content/docs/patch-signing.mdx (hunk 14 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:107]: Ha, good catch. The Google-Doc-to-markdown converter I used turns out to not be very good 😬

Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:107]: Fixed

File: src/content/docs/patch-signing.mdx (hunk 15 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx:115]: Fixed

File: src/content/docs/patch-signing.mdx (hunk 16 above)
Comment (bryanoltman) [reply src/content/docs/patch-signing.mdx]: Fixed

**Pull request merged**